libs/server/Lua/LuaRunner.Functions.cs

```diff
@@ -2773,6 +2774,813 @@ static bool TryDecodeLargeMap(LuaRunner self, ref ReadOnlySpan<byte> data, out i
+                }
+                else
+                {
+                    // Endian mismatch: pick the opposite method to get the bytes in header's endian format
+                    if (BitConverter.IsLittleEndian)
+                        BinaryPrimitives.WriteDoubleBigEndian(into[0..], value);
+                    else
+                        BinaryPrimitives.WriteDoubleLittleEndian(into[0..], value);
+                }
+
+                self.scratchBufferBuilder.MoveOffset(8);
+                self.state.Remove(stackIndex);
+
+                constStrRegisteryIndex = -1;
+                return true;
+            }
+            
+            static bool TryEncodeBytes(LuaRunner self, int stackIndex, char opt, ref int size, out int constStrRegisteryIndex)
+            {
+                Debug.Assert(self.state.Type(stackIndex) == LuaType.String, "Expected string");
+
+                self.state.KnownStringToBuffer(stackIndex, out var data);
+
+                int l = data.Length;
+
+                if (size == 0)
+                {
+                    size = l;
+                }
+
+                if (l < size)
+                {
+                    // string too short, expected at least {size} bytes but got {l}
+                    constStrRegisteryIndex = self.LuaWrappedError(1, self.constStrs.ErrBadArg);
+                    return false;
+                }
+
+                var into = self.scratchBufferBuilder.ViewRemainingArgSlice(data.Length).Span;
+                data.CopyTo(into[0..]);
+                self.scratchBufferBuilder.MoveOffset(data.Length);
+
+                if (opt is 's')
+                {
+                    AddAlignmentPadding(self);
+                    size++;
+                }
+
+                self.state.Remove(stackIndex);
+
+                constStrRegisteryIndex = -1;
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Entry point for struct.unpack from a Lua script.
+        /// </summary>
+        internal int StructUnpack(nint luaStatePtr)
+        {
+            state.CallFromLuaEntered(luaStatePtr);
+
+            var numLuaArgs = state.StackTop;
+
+            if (numLuaArgs == 0 || !TryGetFormat(this, 1, out var format))
+            {
+                return LuaWrappedError(0, constStrs.BadArgPack);
+            }
+
+            // Use stackIndex 1 become when try get format, it also removes format from the lus stack
+            state.KnownStringToBuffer(1, out var data);
+
+            int pos = 0;
+            if (numLuaArgs >= 3) // Only check if the 3rd argument is passed
+            {
+                if (state.Type(3) != LuaType.Number || state.CheckNumber(3) <= 0)
+                {
+                    return LuaWrappedError(0, constStrs.BadArgPack);
+                }
+                pos = (int)state.CheckNumber(3);
+                pos--;
+                // slice data to start from pos
+                data = data[pos..];
```

Comment:
We need a range check here, throwing an exception is really bad from Lua.

Reply:
the data is truncated is removed. I use a copy slice instead so it's less confusing.